pkgs/build-support/node/fetch-yarn-deps/default.nix

```diff
@@ -131,23 +134,68 @@ in
             dontUnpack = src == null;
             dontInstall = true;
 
-            nativeBuildInputs = [
-              prefetch-yarn-deps
-              cacert
-            ];
+            nativeBuildInputs =
+              [ cacert ]
+              ++ (
+                if (yarnVersion == 1) then
+                  [ prefetch-yarn-deps ]
+                else if (yarnVersion == 3) then
+                  [ yarn-berry3 ]
+                else if (yarnVersion == 4) then
+                  [ yarn-berry4 ]
+                else
+                  [ ]
+              );
             GIT_SSL_CAINFO = "${cacert}/etc/ssl/certs/ca-bundle.crt";
             NODE_EXTRA_CA_CERTS = "${cacert}/etc/ssl/certs/ca-bundle.crt";
 
-            buildPhase = ''
-              runHook preBuild
+            supportedArchitectures = builtins.toJSON {
+              os = [
+                "darwin"
+                "linux"
+              ];
+              cpu = [
+                "arm"
+                "arm64"
+                "ia32"
+                "x64"
+              ];
+              libc = [
+                "glibc"
+                "musl"
+              ];
+            };
```

Comment:
It'd be slightly better if changing this variable would not trigger a change of also the `yarnVersion == 1` derivations. It can be achieved by defining this attribute in the passthru, and using `lib.escapeShellArg`.

Reply:
I don't believe this will change at all.. @szlend this is based on your comment. Do you beliefe this has the potential to change in the foreseeable future?

Reply:
I'm not in the position to merge PRs, but personally I wouldn't touch any of the old builders and just start out fresh, taking pnpm as inspiration. Yarn Berry is a completely different package manager with a different API so imo it doesn't make sense to mix it with the old one.

Reply:
I like that it is behind a unified `fetchYarnDeps`. I first implemented a `fetchYarnDepsBerry` separate function but it 1. looks ugly, 2. doesn't follow the [POLS](https://de.wikipedia.org/wiki/Principle_of_Least_Surprise) 
Although `yarn` and `yarn-berry` are different, they are (IMHO) nicely separated by `yarnVersion`

The API (so, basically the CLI) is nicely hidden behind `fetchYarnDeps` and the new `yarnBerry3ConfigHook`. 

The question here is just, whether the `supportedArchitectures` will change in the future? 
Also, when I ran `nixpkgs-review` on this PR, it did not trigger any mass rebuilds, but only rebuilt a handful of packages. So I believe, even if `supportedArchitectures` would change, it will not trigger a mass rebuild for `v1` packages

Reply:
This a fixed-output derivation so the only way to make it rebuild is to change the hash, or forcefully with --rebuild. If supportedArchitectures changed, nothing would actually happen because the deelrivation with this output has was already in the nix store But it would silently make the derivation non-reproducible.

If this value ever needs to change then we need scan nixpkgs and update the output hashes.

We could predefine all possible supported platforms, or just the one nixpkgs currently supports. Not sure what the best option is here.

Sorry I'm afk this week, writing this on my phone so I can't really have a look at this in any depth until I'm back.

Reply:
This probably won't be an issue:

1. The tests for `fetchYarndeps` still passed. So, although I added the `yarnVersion` variable, I did not have to change their hashes
2. I changed the hash of an unrelated `fetchYarnDeps` from an unrelated package to `lib.fakeHash`, confirmed its the same as before after redownloading all packages, and changed the `supportedArchitectures`. The hash after that was still the same as it was initially. 

So changing`supportedArchitectures` in the future will have no effect on `yarnVersion = 1` packages

Reply:
Yea you are right... I forgot this is a FOD. Marking conversation as resolved.